### PR TITLE
Add transformation function between bases in matrix-free

### PR DESCRIPTION
**Pull request by kronbichler**
The new function provides a fast implementation of the projection from values in quadrature points to the values in the basis for the special case that the number of quadrature points is the same as the number of basis functions.

This projection operation can be interpreted as a change of basis (from the Lagrange polynomials defined in the quadrature points to the actual basis). This operation is sometimes useful in DG methods where a fast change between a representation in quadrature points and in the basis is necessary, e.g. in so-called arbitrary-derivative methods as described in https://doi.org/10.1007/s10915-018-0649-2

### CODE REVIEW

File: include/deal.II/matrix_free/operators.h [hunk 1]
```diff
@@ -999,6 +1034,49 @@ namespace MatrixFreeOperators
...
+                inverse_shape);
+
+    for (unsigned int d = 0; d < n_actual_components; ++d)
+      {
+        const VectorizedArray<Number> *in  = in_array + d * dofs_per_cell;
+        VectorizedArray<Number> *      out = out_array + d * dofs_per_cell;
+
+        if (dim == 3)
+          {
+            evaluator.template hessians<2, true, false>(in, out);
```
Comment (Rombur) [inline include/deal.II/matrix_free/operators.h]: So you just need to apply the Hessian to change basis?

Comment (kronbichler) [reply include/deal.II/matrix_free/operators.h]: It is not really the Hessian, it is a transformation matrix given by the inverse of the original `values` matrix. The Hessian is simply the slot I chose for the operations. This reminds me that I should update the names of these functions at some point. Let me open an issue for that.

File: include/deal.II/matrix_free/operators.h [hunk 2]
```diff
@@ -636,6 +636,41 @@ namespace MatrixFreeOperators
           const VectorizedArray<Number> *               in_array,
           VectorizedArray<Number> *                     out_array) const;
 
+    /**
+     * Transforms from data specified in quadrature points to data in the
+     * original basis. This interprets the data in quadrature points by the
+     * Lagrangian interpolation in the quadrature points and transforms it into
+     * the representation by the basis underlying the current `fe_eval` object.
```
[the remark this answers is not in the document]
Comment (kronbichler) [reply include/deal.II/matrix_free/operators.h]: Fixed.

File: include/deal.II/matrix_free/operators.h [hunk 3]
```diff
@@ -636,6 +636,44 @@ namespace MatrixFreeOperators
...
+     * This operation performs a projection from the data given in quadrature
+     * points to the actual basis underlying this object. This projection can
+     * also be interpreted as a change of the basis from the Lagrange
+     * interpolation polynomials in the quadrature points and to the
+     * representation by the basis underlying the current `fe_eval` object.
+     *
+     * Calling this function on an array as
+     * @code
+     * inverse_mass.transform_from_q_points_to_basis(1, array,
+     * phi.begin_dof_values());
```
Comment (masterleinad) [inline include/deal.II/matrix_free/operators.h]: Can you align this parameter with the others?

File: include/deal.II/matrix_free/operators.h [hunk 4]
```diff
@@ -636,6 +636,44 @@ namespace MatrixFreeOperators
           const VectorizedArray<Number> *               in_array,
           VectorizedArray<Number> *                     out_array) const;
 
+    /**
+     * This operation performs a projection from the data given in quadrature
+     * points to the actual basis underlying this object. This projection can
+     * also be interpreted as a change of the basis from the Lagrange
+     * interpolation polynomials in the quadrature points and to the
+     * representation by the basis underlying the current `fe_eval` object.
```
Comment (masterleinad) [inline include/deal.II/matrix_free/operators.h]: I am not sure that I understand this sentence correctly. Do you want to say
```
[...] as a change of the basis from the Lagrange
interpolation polynomials in the quadrature points to the
representation in the basis underlying the current `fe_eval` object.
```
?

File: include/deal.II/matrix_free/operators.h [hunk 5]
```diff
@@ -636,6 +636,44 @@ namespace MatrixFreeOperators
...
+     *
+     * Calling this function on an array as
+     * @code
+     * inverse_mass.transform_from_q_points_to_basis(1, array,
+     * phi.begin_dof_values());
+     * @endcode
+     * is equivalent to calling
```
Comment (masterleinad) [inline include/deal.II/matrix_free/operators.h]: you can omit "calling"

File: tests/matrix_free/inverse_mass_04.cc [hunk 6]
```diff
@@ -0,0 +1,112 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2014 - 2017 by the deal.II authors
```
Comment (masterleinad) [inline tests/matrix_free/inverse_mass_04.cc]: That should just be 2018.

File: tests/matrix_free/inverse_mass_04.cc [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+#include <deal.II/matrix_free/fe_evaluation.h>
+#include <deal.II/matrix_free/matrix_free.h>
+#include <deal.II/matrix_free/operators.h>
+
+#include <deal.II/numerics/vector_tools.h>
+
+#include "../tests.h"
+
+
+std::ofstream logfile("output");
```
Comment (masterleinad) [inline tests/matrix_free/inverse_mass_04.cc]: Let's get rid off all these `logfile` global variables!

File: tests/matrix_free/inverse_mass_04.cc [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+  deallog << std::endl;
+}
+
+
+
+int
+main()
+{
+  deallog.attach(logfile);
```
Comment (masterleinad) [inline tests/matrix_free/inverse_mass_04.cc]: Just calling `initlog()` should be sufficient.

File: include/deal.II/matrix_free/operators.h (hunk 4 above)
Comment (kronbichler) [reply include/deal.II/matrix_free/operators.h]: Yes, but I will write it as 
```
as a change of the basis from the Lagrange
interpolation polynomials in the quadrature points to the
basis underlying the current `fe_eval` object.
```

File: tests/matrix_free/inverse_mass_04.cc (hunk 7 above)
Comment (kronbichler) [reply tests/matrix_free/inverse_mass_04.cc]: Copy-paste issue. I will remove it from the source test file as well.

**Comment by kronbichler** [comment]
I fixed all issues.

**Comment by bangerth** [comment]
@kronbichler : I'm obviously coming rather late to this PR. Out of curiosity, is the function you have essentially equivalent to `FETools::compute_projection_from_quadrature_points_matrix()`? If so, would it be worth pointing this out in the documentation?

**Pull request merged**